Sources/ScipioKit/Producer/FrameworkProducer.swift

```diff
@@ -194,11 +189,58 @@ struct FrameworkProducer {
         return validFrameworks
     }
 
-    private func restoreAllAvailableCaches(
+    private func restoreAllAvailableCachesIfNeeded(
         availableTargets: Set<CacheSystem.CacheTarget>,
+        to storages: [any CacheStorage],
+        cacheSystem: CacheSystem
+    ) async -> Set<CacheSystem.CacheTarget> {
+        var remainingTargets = availableTargets
+        var restored: Set<CacheSystem.CacheTarget> = []
+
+        for index in storages.indices {
+            let storage = storages[index]
+
+            let logSuffix = "[\(index)] \(storage.displayName)"
+            if index == storages.startIndex {
+                logger.info(
+                    "▶️ Starting restoration with cache storage: \(logSuffix)",
+                    metadata: .color(.green)
+                )
+            } else {
+                logger.info(
+                    "⏭️ Falling back to next cache storage: \(logSuffix)",
+                    metadata: .color(.green)
+                )
+            }
+
+            let restoredPerStorage = await restoreCaches(
+                for: remainingTargets,
+                from: storage,
+                cacheSystem: cacheSystem
+            )
+            restored.formUnion(restoredPerStorage)
+
+            logger.info(
+                "⏸️ Restoration finished with cache storage: \(logSuffix)",
+                metadata: .color(.green)
+            )
+
+            remainingTargets.subtract(restoredPerStorage)
+            if remainingTargets.isEmpty {
```

Comment:
it's kindly to comment this

```suggestion
            // if all frameworks are successfully restored, we don't need to proceed to next cache storage.
            if remainingTargets.isEmpty {
```